src/game.cpp

```diff
@@ -11602,14 +11602,83 @@ void game::read()
     auto filter = []( const item &it ) {
         return it.is_book();
     };
-    int pos = inv_for_filter( _("Read:"), filter );
 
-    if (pos == INT_MIN) {
-        add_msg(_("Never mind."));
+    auto item_loc = inv_map_splice( filter, _("Read:") );
+
+    const int inv_pos = item_loc.get_inventory_position();
+
+    if ( inv_pos != INT_MIN ) {
+        u.reading_from_inventory = true;
+        draw();
+        u.read( inv_pos );
         return;
     }
-    draw();
-    u.read(pos);
+
+    u.reading_from_inventory = false;
+
+    item *it = item_loc.get_item();
+    if ( it == nullptr ) {
+        add_msg( _("Never mind.") );
+        return;
+    }
+
+
+    // check if in vehicle as item cannot be in inventory since the above check "inv_pos != INT_MIN" already handle item in inventory
+    // at the moment, only item_location::item_on_person() will set values returned for get_inventory_position(). All others will return INT_MIN
+    bool item_on_veh = false;
+
+    int part = -1;
+    int item_index = 0;
+    vehicle *veh = m.veh_at( u.pos(), part );
+    point veh_pt( INT_MIN, INT_MIN );
```

Comment:
> I tried using existing functions for picking up items and dropping items so that there will be consistent checking so that if next time there is changes to books pickup/dropping, hopefully I do not have to rewrite the codes. Thus if there is addition of another new condition - e.g. can_pickup_books assuming there is a mutation that will change hands into paws, there will be a chance of not needing to rewrite the codes for this part. Also, will not forget to add in the new condition in this part.

`Character::i_add` is also an existing function. Btw. your version does not check `player::can_pickup`, which checks whether the character is wielding an item that prevents picking things up. This is not done in `Pickup::do_pickup`, but in `Pickup::pick_up`.

I'm all for reusing existing functions, but the added complexity of ~50 lines, *just* to call that function, is not worth it. Btw. the local variable `veh_pt` is written, but never read. This is what I mean with added complexity. And the loops over the item stacks are quite inefficient, item stacks are lists, accessing their content via the array operator is the most inefficient thing to do.